src/pip/_internal/cli/req_command.py

```diff
@@ -158,6 +160,32 @@ def handle_pip_version_check(self, options):
 ]
 
 
+def warn_if_run_as_root():
+    # type: () -> None
+    """Output a warning for sudo users on Unix.
+
+    In a virtual environment, sudo pip still writes to virtualenv.
+    On Windows, users may run pip as Administrator without issues.
+    This warning only applies to Unix root users outside of virtualenv.
+    """
+    if running_under_virtualenv():
+        return
+    if not hasattr(os, "getuid"):
+        return
+    # On Windows, there are no "system managed" Python packages. Installing as
+    # Administrator via pip is the correct way of updating system environments.
+    if sys.platform == "win32" or sys.platform == "cygwin":
+        return
+    if sys.platform == "darwin" or sys.platform == "linux":
+        if os.getuid() != 0:
+            return
+    logger.warning("Running pip as root is dangerous. "
+                   "It is possible to break packages installed for your "
+                   "operating system or by different package managers. "
+                   "You should use a virtual environment rather than trying "
+                   "to run as root. This may help resolve permission issues.")
```

Comment:
I would try making the message shorter. The longer the warning is, the less likely users would read it.

Reply:
This is shorter, yet OTOH the user has no idea what to do instead:

```python
logger.warning("Running pip as root may destroy your system.")`
```

Do you think providing a link would work? I.e.:

```python
logger.warning("Running pip as root may destroy your system. See https://bit.ly/pip-root for details.")`
```

Reply:
A link to the documentation feels like a good compromise to me.

Reply:
There's two situations I can think of where someone would run pip with sudo:

1. Unintentionally: it's often easy to forget which commands need sudo and which don't. In this case, the "Running pip as root may break packages" part of the message should be enough to dissuade them; I don't think we'll need to link an essay.
2. Intentionally: A user may run pip *without* sudo, encounter a permission error, and retry with sudo. We should discourage this by suggesting virtual environments. This is where a link could help imo.

How's this?
```
Running pip as root may break packages and permissions. Permission errors should be fixed with venv: https://docs.python.org/3/library/venv.html 
```

Reply:
Either of the followings would be a better link IMO:

* https://packaging.python.org/tutorials/installing-packages/#creating-virtual-environments
* https://docs.python.org/3/tutorial/venv.html

The library documentation page covers more on the programmatic interface and isn’t very useful for the audience we target here.

Reply:
I like @hroncok's link-based suggestion. I'd say use `https://pip.pypa.io/warnings/root-usage` for that right now. We can setup a redirect in our RTD configuration, to make it point to the right place once this is merged.

It'll also defer the discussion on what the right place is, which I like. 🙃

Reply:
I still think we should explicitly call out permission errors and venv in the warning message, since that's by far the most common misuse of sudo.

So instead of a generic `https://pip.pypa.io/warnings/root-usage` redirect, how about we specify `https://pip.pypa.io/warnings/venv` and tentatively redirect it to `https://docs.python.org/3/tutorial/venv.html`?

Updated warning message:
`Running pip as root may break packages and permissions. Permission errors should be fixed with venv: https://pip.pypa.io/warnings/venv`

Reply:
How about "Permission errors should be fixed with venv" -> "If you need to install packages not supplied by your system package manager, use a virtual environment: https://pip.pypa.io/warnings/venv"? Focus on what the user wants to do, rather than simply on fixing the error they got.

Reply:
I think we should call this out for what they're doing here, and provide better guidance on what they should be doing there (using a virtualenv, why, what the risks are etc).

Reply:
Shorter message with a focus on the user:
```
    logger.warning("Running pip as root may break packages and permissions. "                       
                   "You should install packages reliably by using venv: "                                             
                   "https://pip.pypa.io/warnings/venv") 
```

I agree that technical details belong in the link.